pravega/pravega
# Issue 4509: Rack-aware Bookkeeper client config

**Change log description**  
Allow the Bookkeeper client in the Segment Store to be configured with rack-aware ensemble placement policy.

**Purpose of the change**  
Fixes #4509.

**What the code does**  
This PR enables the Bookkeeper client in the Segment Store to be configured with a rack-aware ensemble placement policy. To this end, we introduce 3 new Bookkeeper client parameters:
- `bookkeeper.enforceMinNumRacksPerWriteQuorum`: This parameter defines whether we should enforce the rack-aware policy or not.
- `bookkeeper.minNumRacksPerWriteQuorum`: In the case we enforce rack-aware ensamble placement policy, this parameter determines the number of different physical location that a write should be replicated to.
- `bookkeeper.networkTopologyScriptFileName`: This is an important and environment-dependent parameter. It should point to a script that, based on a list of Bookie IPs as input, outputs a list of equal size with the physical locations of each Bookie IP (e.g. `[IP1, IP2, IP3]` -> `[/rack1, /rack2, /rack3]`). We provide a sample file that is intended to be used in test or demo situations only.

**How to verify it**  
We have deployed Pravega with the changes of this branch. As can be see, the `RackawareEnsemblePlacementPolicy` is properly initialized:
```
2020-01-22 18:29:00,845 960  [main] INFO  o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Initialize rackaware ensemble placement policy @ <Bookie:172.28.13.11:0> @ /datacenter/rack-11 : org.apache.bookkeeper.net.ScriptBasedMapping.
```
This means that the Bookkeeper client has correctly initialized the `ScriptBasedMapping`, which internally invokes the topology script (`networkTopologyScriptFileName`).

Then, upon each Bookie join, the network topology is updated correctly, which implies that the script is correctly invoked:
``` 
2020-01-22 18:29:00,873 988  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : left bookies are [], joined bookies are [pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181, pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181], while dead bookies are [].
2020-01-22 18:29:00,877 992  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,877 992  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,878 993  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,881 996  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] INFO  o.a.b.net.NetworkTopologyImpl - Adding a new node: /datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : bookie pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181 joined the cluster.
2020-01-22 18:29:00,885 1000 [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.c.RackawareEnsemblePlacementPolicyImpl - Cluster changed : left bookies are [], joined bookies are [], while dead bookies are [].
```
As you can see, the final topology relates each Bookie to a different rack number, that corresponds to the last number of the Bookie IP (as explained in the sample Bookkeeper topology script):
```
pravega-bookie-0   1/1     Running   1          13m     xxx.8
pravega-bookie-1   1/1     Running   0          13m     yyy.10
pravega-bookie-2   1/1     Running   0          13m     zzz.9
```
```
2020-01-22 18:29:00,884 999  [BookKeeperClientScheduler-OrderedScheduler-0-0] DEBUG o.a.b.net.NetworkTopologyImpl - NetworkTopology became:
/datacenter/rack-10/pravega-bookie-1.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-9/pravega-bookie-2.pravega-bookie-headless.default.svc.cluster.local:3181
/datacenter/rack-8/pravega-bookie-0.pravega-bookie-headless.default.svc.cluster.local:3181
```
